easybuild/easyblocks/f/fftw.py

```diff
@@ -0,0 +1,198 @@
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools.config import build_option
+from easybuild.tools.systemtools import X86_64, get_cpu_features
+from easybuild.tools.toolchain.compiler import OPTARCH_GENERIC
+
+
+# AVX*, FMA, SSE2 (x86_64 only)
+FFTW_CPU_FEATURE_FLAGS_SINGLE_DOUBLE = ['avx', 'avx2', 'avx512', 'fma', 'sse2', 'vsx']
+# Altivec (POWER), SSE (x86), NEON (ARM), FMA (x86_64)
+# asimd is CPU feature for extended NEON on AARCH64
+FFTW_CPU_FEATURE_FLAGS = FFTW_CPU_FEATURE_FLAGS_SINGLE_DOUBLE + ['altivec', 'asimd', 'neon', 'sse']
+FFTW_PRECISION_FLAGS = ['single', 'double', 'long-double', 'quad-precision']
+
+
+class EB_FFTW(ConfigureMake):
+    """Support for building/installing FFTW."""
+
+    @staticmethod
+    def _prec_param(prec):
+        """Determine parameter name for specified precision"""
+        return 'with_%s_prec' % prec.replace('-', '_').replace('_precision', '')
+
+    @staticmethod
+    def extra_options():
+        """Custom easyconfig parameters for FFTW."""
+        extra_vars = {
+            'auto_detect_cpu_features': [True, "Auto-detect available CPU features, and configure accordingly", CUSTOM],
+            'with_mpi': [True, "Enable building of FFTW MPI library", CUSTOM],
+            'with_openmp': [True, "Enable building of FFTW OpenMP library", CUSTOM],
+            'with_threads': [True, "Enable building of FFTW threads library", CUSTOM],
+        }
+
+        for flag in FFTW_CPU_FEATURE_FLAGS:
+            help_msg = "Configure with --enable-%s (if None, auto-detect support for %s)" % (flag, flag.upper())
+            extra_vars['use_%s' % flag] = [None, help_msg, CUSTOM]
+
+        for prec in FFTW_PRECISION_FLAGS:
+            help_msg = "Enable building of %s precision library" % prec.replace('-precision', '')
+            extra_vars[EB_FFTW._prec_param(prec)] = [True, help_msg, CUSTOM]
+
+        return ConfigureMake.extra_options(extra_vars)
+
+    def __init__(self, *args, **kwargs):
+        """Initialisation of custom class variables for FFTW."""
+        super(EB_FFTW, self).__init__(*args, **kwargs)
+
+        for flag in FFTW_CPU_FEATURE_FLAGS:
+            # fail-safe: make sure we're not overwriting an existing attribute (could lead to weird bugs if we do)
+            if hasattr(self, flag):
+                raise EasyBuildError("EasyBlock attribute '%s' already exists")
+            setattr(self, flag, self.cfg['use_%s' % flag])
+
+        cpu_features = get_cpu_features()
+
+        # on macOS, AVX is indicated with 'avx1.0' rather than 'avx'
+        if 'avx1.0' in cpu_features:
+            cpu_features.append('avx')
+
+        self.log.info("List of CPU features: %s", cpu_features)
+
+        # auto-detect CPU features that can be used and are not enabled/disabled explicitly,
+        # but only if --optarch=GENERIC is not being used
+        if self.cfg['auto_detect_cpu_features'] and build_option('optarch') != OPTARCH_GENERIC:
```

Comment:
All x86_64 CPUs support at least SSE2, so this could be enabled even with `OPTARCH_GENERIC`.